Mathlib/Analysis/PSeries.lean

```diff
@@ -19,134 +19,180 @@ converges if and only if so does `∑ k, 2 ^ k f (2 ^ k)`. We prove this test in
 `NNReal.summable_condensed_iff` and `summable_condensed_iff_of_nonneg`, then use it to prove
 `summable_one_div_rpow`. After this transformation, a `p`-series turns into a geometric series.
 
-## TODO
-
-It should be easy to generalize arguments to Schlömilch's generalization of the Cauchy condensation
-test once we need it.
-
 ## Tags
 
 p-series, Cauchy condensation test
 -/
 
 /-!
-### Cauchy condensation test
+### Schlömilch's generalization of the Cauchy condensation test
 
-In this section we prove the Cauchy condensation test: for `f : ℕ → ℝ≥0` or `f : ℕ → ℝ`,
-`∑ k, f k` converges if and only if so does `∑ k, 2 ^ k f (2 ^ k)`. Instead of giving a monolithic
-proof, we split it into a series of lemmas with explicit estimates of partial sums of each series in
-terms of the partial sums of the other series.
+In this section we prove the Schlömilch's generalization of the Cauchy condensation test:
+for `u : ℕ → ℕ` and `f : ℕ → ℝ≥0` or `f : ℕ → ℝ`, `∑ k, f k` converges if and only if
+so does `∑ k, (u (k + 1) - u k) * f (u k)`. Instead of giving a monolithic proof, we split it
+into a series of lemmas with explicit estimates of partial sums of each series in terms of the
+partial sums of the other series.
 -/
 
+/--
+A sequence `u` has the property that its ratio of successive differences is bounded
+when there is a positive real number `C` such that, for all n ∈ ℕ,
+(u (n + 2) - u (n + 1)) ≤ C * (u (n + 1) - u n)
+-/
+def SuccDiffBounded (C : ℕ) (u : ℕ → ℕ) : Prop :=
+  ∀ n : ℕ, u (n + 2) - u (n + 1) ≤ C • (u (n + 1) - u n)
 
 namespace Finset
 
 open BigOperators
 
-variable {M : Type*} [OrderedAddCommMonoid M] {f : ℕ → M}
+variable {M : Type*} [OrderedAddCommMonoid M] {f : ℕ → M} {u : ℕ → ℕ}
 
-theorem le_sum_condensed' (hf : ∀ ⦃m n⦄, 0 < m → m ≤ n → f n ≤ f m) (n : ℕ) :
-    (∑ k in Ico 1 (2 ^ n), f k) ≤ ∑ k in range n, 2 ^ k • f (2 ^ k) := by
+theorem le_sum_schlomilch' (hf : ∀ ⦃m n⦄, 0 < m → m ≤ n → f n ≤ f m) (h_pos : ∀ n, 0 < u n)
+    (hu : Monotone u) (n : ℕ) :
+    (∑ k in Ico (u 0) (u n), f k) ≤ ∑ k in range n, (u (k + 1) - u k) • f (u k) := by
   induction' n with n ihn
   · simp
-  suffices (∑ k in Ico (2 ^ n) (2 ^ (n + 1)), f k) ≤ 2 ^ n • f (2 ^ n) by
+  suffices (∑ k in Ico (u n) (u (n + 1)), f k) ≤ (u (n + 1) - u n) • f (u n) by
     rw [sum_range_succ, ← sum_Ico_consecutive]
     exact add_le_add ihn this
-    exacts [n.one_le_two_pow, Nat.pow_le_pow_of_le_right zero_lt_two n.le_succ]
-  have : ∀ k ∈ Ico (2 ^ n) (2 ^ (n + 1)), f k ≤ f (2 ^ n) := fun k hk =>
-    hf (pow_pos zero_lt_two _) (mem_Ico.mp hk).1
+    exacts [hu n.zero_le, hu n.le_succ]
+  have : ∀ k ∈ Ico (u n) (u (n + 1)), f k ≤ f (u n) := fun k hk =>
+    hf (Nat.succ_le_of_lt (h_pos n)) (mem_Ico.mp hk).1
   convert sum_le_sum this
   simp [pow_succ, two_mul]
-#align finset.le_sum_condensed' Finset.le_sum_condensed'
```

Comment:
You're removing #align-ed declarations: as a heads-up, I think this is undesirable.
(These were created during porting mathlib from Lean 3 to Lean 4. As I understand it, the current plan is to remove them at some point - but it's not clear when that will be.)

I'm not sure what the best fix is: keep the old declarations as a special case? (Sometimes, that's worth it anyway, if the special case is the version many people need.) Point the align to the new version?

 I'll ask on zulip.

Reply:
Thanks for the thread on Zulip. I have added back the theorems `summable_condensed_iff` and `summable_condensed_iff_of_nonneg` since they are more frequently used. I am not sure if we should keep the old declarations for the rest of the theorems since they are really lemmas for the final two theorems.